Writing a readme on using the new build (#79527)
swiftlang/swift: Runtimes/Core/Readme.md

# SwiftCore

SwiftCore contains core libraries that sit below the platform overlays.
These include the standard library and associated runtimes, SwiftOnoneSupport,
and the concurrency runtimes.

## Build Instructions

> [!IMPORTANT]
> The standard library requires that it is built with a Swift compiler that is
> at least as new as the standard library sources. You will likely need to
> build the compiler as you would normally.
> In these instructions, `<swiftc>` is the path to your just-built Swift
> compiler.

> First create a build directory in the `Core/` directory and navigate to it.
> 
> ```sh
> mkdir build && cd build
> ```
> 
> Configure a simple default build. This results in a static build of the standard
> library.
> 
> ```sh
> cmake -G 'Ninja' -DCMAKE_Swift_COMPILER=<swiftc> ../`
> ninja
> ```

Reviewer: ```suggestion
Configure a simple default build, which builds a static version of the standard library.

```sh
cmake -B build -G Ninja -S . -D CMAKE_Swift_COMPILER=<swiftc>
ninja -C build
```
```

I'd like to throw in some bike shedding for fun. I tend to prefer the build directory to be named `out`.

To build a dynamic library, pass `-DBUILD_SHARED_LIBS=YES` to CMake.

```sh

> cmake . -DBUILD_SHARED_LIBS=YES
> ninja

Reviewer: ```suggestion
cmake -B build -G Ninja -S . -D CMAKE_Swift_COMPILER=<swiftc> -D BUILD_SHARED_LIBS=YES
ninja -C build
```

This keeps the build rules nearly identical and clearly shows the difference.

Author: I don't think I agree. Once you have a build configured, we only need to show what needs to change in order to get the desired effect. The others are just clutter.

```

To install the standard library to a staging directory, set the `DESTDIR`
environment variable and run `ninja install`

```sh

> DESTDIR=SwiftCore-Staging ninja install

Reviewer: I don't think that this is portable. The environment variable handling is not identical on all platforms. If we want to do this, we should encourage the use of `cmake -E env DESTDIR=... ...`

```

Semantic editing support is nice during development. CMake 3.29 and newer can
produce `compile_commands.json` compilation databases, which tools like
SourceKit-LSP and clangd use to provide a semantic editing experience with
features like semantic syntax highlighting, symbol hover information, and
jump-to-definition.

```sh

> cmake . -DCMAKE_EXPORT_COMPILE_COMMANDS=YES

Reviewer: ```suggestion
cmake -B build -G Ninja -S . -D CMAKE_Swift_COMPILER=<swiftc> -D CMAKE_EXPORT_COMPILE_COMMANDS=YES
```

```

> [!NOTE]
> Some editors will look under `<source-root>/build` for the generated
> `compile_commands.json` file while others do not. You may want to create a
> symlink from the `Core/` directory to `build/compile_commands.json`.

> > `ln -s build/compile_commands.json`

Reviewer: Again, this command is not portable. I think that we should encourage the use of `cmake -E create_symlink` if we want to have a copy-pastable command.

> The compile-commands are specific to the current build configuration, so the

Reviewer: ```suggestion
The compile commands are specific to the current build configuration, so the
```

Author: compile modifies command. It's not a command telling a dog to sit, it's a command for compiling.

semantic results shown in the editor match what you are currently building.

There are many more knobs for configuration. From the build directory, run
`ccmake` to edit the build configuration to your liking.

### Reproducing a Build

The knobs make it easy to build the standard library to your needs, but can make
it challenging to reproduce a specific build configuration.
CMake has cache files to specify how to position the knobs for a given
configuration.

These caches live under `cmake/caches`, but can live anywhere.

> To reproduce the Apple intel macOS standard library build, one would use the

Reviewer: ```suggestion
To reproduce the Apple Intel macOS standard library build, one would use the
```

following CMake command from the build directory:

```sh

> cmake -G 'Ninja' \
>   -DCMAKE_OSX_DEPLOYMENT_TARGET=15.3 \
>   -DCMAKE_OSX_SYSROOT=macosx \
>   -DCMAKE_Swift_COMPILER=<swiftc> \
>   --toolchain ../cmake/caches/Vendors/Apple/Darwin.toolchain.cmake \
>   -C ../cmake/cache/Vendors/Apple/x86_64-MacOSX.cmake \
>   ../
> ninja

Reviewer: ```suggestion
cmake -B build -G Ninja -S . \
  -D CMAKE_OSX_DEPLOYMENT_TARGET=15.3 \
  -D CMAKE_OSX_SYSROOT=macosx \
  -D CMAKE_Swift_COMPILER=<swiftc> \
  --toolchain cmake/caches/Vendors/Apple/Darwin.toolchain.cmake \
  -C cmake/cache/Vendors/Apple/x86_64-MacOSX.cmake
ninja -C build
```
